Clean up form 'breadcrumb' paths

cooperka opened the pull request:
Cleans up the logic to generate 'breadcrumb' paths in the jump screen to make the path more understandable to users.

<img width="305" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51272664-b9890980-1998-11e9-84b5-dfc12750071b.png">

Also updates the item names in the repeat picker in a similar way.

<img width="305" alt="screenshot" src="https://user-images.githubusercontent.com/2047062/51273028-acb8e580-1999-11e9-9b1b-2fb992fe1dea.png">

#### Examples

e.g. A -- normal, unlabeled groups:

| Screen | Old path | New path |
|---|---|---|
| Friends repeat instance list (jump screen) | Friends | Friends |
| Alice (jump screen) | Friends (1) | Friends > 1 |
| Question about Alice (question view) | Friends (1) | Friends > 1 |

e.g. B -- labeled groups:

| Screen | Old path | New path |
|---|---|---|
| Friends repeat instance list (jump screen) | Friends | Friends |
| Alice (jump screen) | Friends (1) | Friends > 1 > Alice |
| Question about Alice (question view) | Friends (1) > Alice | Friends > 1 > Alice |

e.g. C -- deeper labeled groups:

| Screen | Old path | New path |
|---|---|---|
| Pets repeat instance list (jump screen) | Friends (1) > Pets | Friends > 1 > Alice > Pets |
| Dog (jump screen) | Friends (1) > Alice > Pets (2) | Friends > 1 > Alice > Pets > 2 > Dog |
| Question about Dog (question view) | Friends (1) > Alice > Pets (2) > Dog | Friends > 1 > Alice > Pets > 2 > Dog |

e.g. D -- repeat picker items:

| Group type | Old name | New name |
|---|---|---|
| Unlabeled | Friends (1) | Friends > 1 |
| Labeled | Alice (1) | 1. Alice |

#### What has been done to verify that this works as intended?

Extensive manual testing with:
- Friends ([XML](https://drive.google.com/open?id=16DBIr5-U7DUY0_KN3tbhUS4R5WYICwIG)) (unlabeled group)
- nested-repeats-complex ([XML](https://drive.google.com/open?id=1zOvDEZz6vwQr5IBT91k7MuDMZ-hJDats) / [XLS](https://drive.google.com/open?id=1xVv682pbBMQ_MTZdNnj6aP11jhPp6fT8WQ42oEr7WaA)) (labeled, top-level "friends" and "enemies" groups; nested, labeled "pets" group)

Navigating in and out of nested repeat groups, adding children, editing names, using the back button, jumping back and forth between the FormEditor and FormHierarchy views.

#### Why is this the best possible solution? Were any other approaches considered?

This solution makes it clear to users where they are in the form, and also makes the path more consistent with the names of repeat group instances (e.g. `friends > 1` or `1. Alice`).

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The new paths are more intuitive, but it could potentially disorient existing users who've come to expect a precise string to be displayed (whether from personal experience or from training screenshots).

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

None that I can tell.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

cooperka committed:
Clean up some path-building code

Simpler, but now the form index (1) appears while in the repeat group
picker when it shouldn't. This will be addressed shortly.

cooperka committed:
Only move path index back if on a question

Seems to be an old bug (since ~2012). This makes it so that the current
group is shown in the path as expected.

cooperka committed:
Postpone the path multiplicity until the next group

Show it on the child instance, not the parent group.

grzesiek2010 commented:
@cooperka 

I've noticed a bug in your solution. Everything looks fine if you set group labels in the way you did that in your sample form but I build a form using ODK Build and then a group label is simple group name + group number in brackets.

Here is a form: 
[Friends.xml.txt](https://github.com/opendatakit/collect/files/2702858/Friends.xml.txt)

Then on your branch I can't see any group number:

| Mater branch  | This pr |
| ------------- | ------------- |
| ![screenshot_2018-12-21-14-11-38](https://user-images.githubusercontent.com/3276264/50344221-630be500-052a-11e9-87cb-dd4c28667780.png) | ![screenshot_2018-12-21-14-06-40](https://user-images.githubusercontent.com/3276264/50344218-62734e80-052a-11e9-8839-b619841f05da.png)  |

zestyping commented:
The implementation here looks entirely reasonable, assuming we agree on what we want to see in the UI.  I have some thoughts about the UI design, though.

1. This design replaces **Friends (1) > Pets** with **Friends > Alice (1) > Pets**.  I agree that inserting "Alice" is an improvement in clarity.

2. This design replaces **Friends (1) > Alice** with **Friends > Alice (1)**.  To me, putting the "(1)" after "Alice" is more confusing.  The part that's plural is "Friends", so I expect the number to be next to "Friends", or at least between "Friends" and "Alice" because the number describes the relationship between "Friends" and "Alice".

I think a small change would make this clearer in the UI.  If I imagine writing out the structure of the form as an outline list, it would look like this:

```
• Friends
    1. Alice
        • Pets
            1. Dog
            2. Cat
    2. Charlie
        • Pets
           1. Parrot
           2. Gecko
```

So, how about putting the "1" in front of Alice, like this?

**Friends > 1. Alice > Pets**

This is the clearest way I can think of to show that "Alice" is the first item in a series of Friends.

cooperka commented:
Thanks for the bug info @grzesiek2010. I can reproduce; it seems like the reason this happens is that my form has an extra group inside the repeat to show a label:

```xml
<friends jr:template="">
    <friends_label>
        <!-- List of questions -->
```

and yours does not:

```xml
<friends jr:template="">
    <!-- List of questions -->
```

~~I'll add a fallback so that if there's no inner group, it will show the number in parenthesis just like in your `master` screenshot. I don't know of any other way to handle this, since if there's no label, there's no breadcrumb to show in the path (unless we make something up, e.g. "Child" or something).~~

I'm still considering how to handle this. There could be multiple, nested, unlabeled groups... as suggested in https://github.com/opendatakit/collect/pull/2763#pullrequestreview-189581803 I think displaying `Friends > 1` makes sense instead of `Friends (1)`.

cooperka commented:
@zestyping I like your numbering suggestion; I think that makes it even more readable. Presumably the field item titles should also be changed, e.g. this screenshot of the repeat picker should use the new numbering system too:

![repeat picker screenshot](https://user-images.githubusercontent.com/2047062/50915493-ebcdc080-1406-11e9-9361-c000cddb3a92.png)

And whatever solution we end up with to the bug mentioned above will also need to apply here.

Note that I'm currently on holiday until this Monday the 14th but I'll do my best to make time so this can get merged before the feature freeze.

cooperka commented:
What do you both think of this proposal (last row of table):

| | List (labeled) | List (unlabeled) | Path for Alice (labeled) | Path for Alice (unlabeled) |
|---|---|---|---|---|
| **master** | Alice (1)<br />Bob (2) | Friends (1)<br />Friends (2) | Friends (1) | Friends (1) |
| **d3d8537** | Alice (1)<br />Bob (2) | Friends (1)<br />Friends (2) | Friends > Alice (1) | Friends [bug] |
| **[proposed]** | 1. Alice<br />2. Bob | Friends > 1.<br />Friends > 2. | Friends > 1. Alice | Friends > 1. |

This will also need to apply to upcoming visible groups (https://github.com/opendatakit/collect/pull/2764), which can also be labeled/unlabeled. It  seems feasible, just requires looking ahead while building the path to see what the next label is in order to modify the current label.

smoyte commented:
@zestyping I see your point and I like the 1. Alice idea. To me, **Friends (1) > Alice** is a nonstarter because if you imagine there are three friends, their crumbs will look like:

* Friends (1) > Alice
* Friends (2) > Bob
* Friends (3) > Chaz

I have never seen breadcrumbs like this anywhere in my life. This makes it seem like Alice, Bob, and Chaz are each in a separate group, each group is called "Friends", and they're given a numeric suffix to disambiguate them.

@cooperka @zestyping I think the **1. Alice** style kind of breaks down in the unlabeled case. To me, **Friends > 1.** looks strange, like something got deleted.

That is why I kind of liked the (1) suffix because then in the unlabeled case you get **Friends > (1)**, **Friends > (2)**, etc., which looks halfway decent.

We could do:

* Friends > (1) Alice
* Friends > (2) Bob
* Friends > (3) Chaz

which would be 

* Friends > (1)
* Friends > (2)
* Friends > (3)

in the unlabeled case.

Thoughts?

@grzesiek2010 Thanks for catching the bug!

yanokwa commented:
@cooperka I didn't know you were on vacation until the 14th. I think it's important that we get this in for this round, so if we need to extend the freeze date by a few days.

@smoyth I think the `(n)` construction is a bit heavy handed. I agree that having the period looks weird, so what if we drop the period on the unlabeled list/path? And if we really want to kick up a notch, we can use a bit of typography or color (😱) to distinguish the number from the label.

smoyte commented:
@yanokwa Just to be clear, (n) is what is already there. You're saying "heavy handed" in comparison to some new ideal, not the current state of things, right?

I'd be fine with dropping the period. So we'd have

Friends > 1. Alice
Friends > 2. Bob
Friends > 3. Chaz

and 

Friends > 1
Friends > 2
Friends > 3

and in the doubly-nested case,

Friends > 1 > Pets > 1
Friends > 1 > Pets > 2
Friends > 2 > Pets > 1

etc.

Yes? Looks fine to me!

cooperka committed:
Simplify path building; deliminate multiplicity numbers

e.g. instead of `Friends (1)` it will show `Friends > 1` in the path.

cooperka committed:
Number repeat items in front instead of parenthetically

e.g. instead of `Friends (1)` it will show `Friends > 1` or `1. Alice`
for repeat group items.

cooperka commented:
Code + screenshots + description have been updated.

Note that instead of `Friends > 1. Alice` it will show `Friends > 1 > Alice` because the form I've been testing with uses a workaround for dynamically labeling groups. It uses nested groups which results in the extra layer in the path. Based on @yanokwa's comment [here](https://github.com/opendatakit/collect/pull/2764#issuecomment-454219524), I left it like this for simplicity. It seems good to me but feedback is appreciated as always.

@opendatakit-bot label "needs review"

cooperka committed:
Retain NonNull annotation for method wrapper

cooperka committed:
Add back LRM character

To prevent RTL languages from moving the period to the front.

zestyping commented:
Thanks for making this change, @cooperka!

Is the display `Friends > 1 > Alice` due to something unusual about this form, or is it the implementation?

Would the most common way of designing a form for entering data about a list of friends show `Friends > 1. Alice` or `Friends > 1 > Alice`?

(I hope it's `Friends > 1. Alice` because it reads better, and the second implies that `Friends > 1` has a different number of levels than `Friends > 1 > Alice`, which isn't true.)

cooperka commented:
> Is the display Friends > 1 > Alice due to something unusual about this form, or is it the implementation?

It's an unusual form structure in that there's a second group to show the "Alice" label. `Alice` is in fact a sub-group of `Friends > 1`.

> Would the most common way of designing a form for entering data about a list of friends show Friends > 1. Alice or Friends > 1 > Alice?

Yaw suggested in the linked comment that it may be fairly common. This is something we can further improve if it ends up being an issue, but as it stands, I think this PR makes some significant improvements (as well as bug fixes) to the UI.

Many forms will show just `Friends > 1`, with `Alice` never appearing in the path. To be clear, `Friends > 1. Alice` is **not** a possible path right now.

zestyping commented:
Okay, thanks for explaining.

I thought the way it was described in @smoyth's [comment above](https://github.com/opendatakit/collect/pull/2762#issuecomment-453643280) was perfect, and that it was what we were going for.  Do we still like it as an eventual goal?

I do agree that the current PR is an improvement as it stands.

I'm a little confused about one thing—in your original implementation, you had it showing `Friends > Alice (1)` in this case.  What makes `Friends > Alice (1)` possible but `Friends > 1. Alice` impossible?

cooperka commented:
> Do we still like it as an eventual goal?

Yes, especially if we determine that a lot of forms use this workaround. Alternatively there may be some better solution that comes along and people migrate to that instead of the current workaround.

> What makes Friends > Alice (1) possible but Friends > 1. Alice impossible?

Both would be possible with my original implementation, but that leads to a bug that Grzegorz pointed out with "normal" forms. Forms that don't use the workaround would no longer display any item number in the path.

In order to support regular forms **and** forms that use the workaround, we'd need to write logic that specifically detects the workaround and modifies the path to hide one of the groups.

zestyping commented:
Makes sense. Thanks!

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -234,29 +233,33 @@ protected void goUpLevel() {
 
-        String path = ODKView.getGroupsPath(groups.toArray(new FormEntryCaption[groups.size()]));
+        // If the repeat picker is showing, don't show an item number for the current index.
+        boolean hideLastMultiplicity = shouldShowRepeatGroupPicker();
```
I think this variable is redundant. The parameter in `getGroupsPath()` is named `hideLastMultiplicity` so ii think it's clear enough and you can call `shouldShowRepeatGroupPicker()` directly.

cooperka replied:
I separated out this variable in order to explain why `hideLastMultiplicity` is used this way. If it's all squished into the return statement, it makes it less clear what the comment is referring to. I would prefer to keep it like this, do you have a strong preference?

mmarciniak90 commented:
@cooperka I used a form where 3 groups exist : 
1. group1text
2. group2date
3. group3media

[groups.xml.txt](https://github.com/opendatakit/collect/files/2773147/groups.xml.txt)

I added the only 3rd group and I opened `Jump screen`

<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51391138-8ded5700-1b31-11e9-8369-ebee96fe731c.png">

Is it expected that `group1text > 1> group3media` is visible?

cooperka committed:
PR fix: Handle groups at very beginning of form

The very first group in a form is auto-entered, so this steps back out.

cooperka commented:
@mmarciniak90 that issue has now been fixed, can you please verify?

Builds are failing due to:

> A problem occurred configuring project ':collect_app'.
> > Failed to install the following Android SDK packages as some licences have not been accepted.
>      platforms;android-28 Android SDK Platform 28
>   To build this project, accept the SDK license agreements and install the missing components using the Android Studio SDK Manager.
>   Alternatively, to transfer the license agreements from one workstation to another, see http://d.android.com/r/studio-ui/export-licenses.html

yanokwa commented:
@cooperka Don't worry about the CI issue. I'll look into it shortly. Either way, it won't be a blocker to getting this tested. 

Update: You can track the CircleCI issue at https://discuss.circleci.com/t/android-platform-28-sdk-license-not-accepted/27768

mmarciniak90 commented:
@cooperka thanks for such a quick fix!

**Tested with success.**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified forms:
* fewRepeatGroups
* Friends
* groups
* nested-index-repeat
* nested-repeats
* nested-repeats-complex
* nRepeats
* RepeatGroupNew
* RepeatGroupAndQuestion
* repeatTest
* RepeatSimple

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
